app-kotlin/src/main/kotlin/au/com/gridstone/trainingkotlin/views/HomeViewController.kt

```diff
@@ -0,0 +1,53 @@
+package au.com.gridstone.trainingkotlin.views
+
+import android.os.Bundle
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+import android.widget.ImageView
+import android.widget.ProgressBar
+import androidx.core.view.isVisible
+import androidx.recyclerview.widget.LinearLayoutManager
+import androidx.recyclerview.widget.RecyclerView
+import au.com.gridstone.trainingkotlin.R
+import au.com.gridstone.trainingkotlin.adapters.HomeListAdapter
+import au.com.gridstone.trainingkotlin.viewmodels.HomeViewModel
+import au.com.gridstone.trainingkotlin.viewmodels.HomeViewState
+import au.com.gridstone.trainingkotlin.viewmodels.HomeViewState.*
+import com.bluelinelabs.conductor.Controller
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.launch
+import org.koin.core.component.KoinComponent
+import org.koin.core.component.get
+
+class HomeViewController : Controller(), KoinComponent {
+
+  private val viewModel: HomeViewModel = get()
+
+  override fun onCreateView(
+    inflater: LayoutInflater,
+    container: ViewGroup,
+    savedViewState: Bundle?,
+  ): View = inflater.inflate(R.layout.home_view_controller, container, false)
+
+  override fun onAttach(view: View) {
+    val recyclerView: RecyclerView = view.findViewById(R.id.home_recyclerView)
+    val progressBar: ProgressBar = view.findViewById(R.id.home_progressBar)
+    val errorImage: ImageView = view.findViewById(R.id.image_error)
+    val layoutManager = LinearLayoutManager(applicationContext)
+    recyclerView.layoutManager = layoutManager
+    val homeListAdapter = HomeListAdapter(emptyList(), router)
+    recyclerView.adapter = homeListAdapter
+    GlobalScope.launch(Dispatchers.Main) {
```

Comment:
Avoid `GlobalScope`. Instead, scope this launched coroutine to the controller scope

Reply:
thanks @xxfast  well presented and explained cheers.